### PR TITLE
Remove FieldTemplate<> from field reflection type

### PR DESCRIPTION
**Pull request by veselinp**
- FieldTemplate<> params are part of field relfection type, this unnecessary increases  generated symbol names.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -19,6 +19,10 @@ different versioning scheme, following the Haskell community's
 * C# NuGet version: TBD (minor bump needed)
 * C# Comm NuGet version: TBD (minor bump needed)
 
+### C++ & `gbc` ###
+
+* Remove FieldTemplate details from field reflection type.
+
```
Comment (chwarr) [inline CHANGELOG.md]: Minor: there's a `gbc` section right below where this would make more sense. Consider also duplicating the entry in the C++ section as well.

File: CHANGELOG.md [hunk 2]
```diff
@@ -19,6 +19,10 @@ different versioning scheme, following the Haskell community's
 * C# NuGet version: TBD (minor bump needed)
 * C# Comm NuGet version: TBD (minor bump needed)
 
+### C++ & `gbc` ###
+
+* Remove FieldTemplate details from field reflection type.
```
Comment (chwarr) [inline CHANGELOG.md]: Minor: can you expand on this and say why we're making the change (to shorten the symbol names)?

**Comment by veselinp** [comment]
I've pushed a fixup, clang should be happy now.

**Comment by chadwalters** [comment]
clang build is happy. Looks like VC12 is still broken though.

File: CHANGELOG.md (hunk 2 above)
Comment (veselinp) [inline CHANGELOG.md]: Done!

File: CHANGELOG.md (hunk 1 above)
Comment (veselinp) [inline CHANGELOG.md]: Done!

**Comment by veselinp** [comment]
vc12 doesn't understand constexpr :)

**Comment by veselinp** [comment]
now, clang is broken, I'm looking again

File: cpp/test/core/container_extensibility.cpp [hunk 3]
```diff
@@ -123,8 +123,8 @@ TEST_CASE_BEGIN(MultiIndexTest)
                 SimpleStructView, 
                 boost::multi_index::indexed_by<
                     boost::multi_index::sequenced<>,
-                    ordered_non_unique_field<SimpleStructView::Schema::var::m_str>,
-                    ordered_non_unique_field<SimpleStructView::Schema::var::m_uint64>
+                    ordered_non_unique_field<SimpleStructView::Schema::var::m_str, get_field_template<SimpleStructView::Schema::var::m_str>::type>,
+                    ordered_non_unique_field<SimpleStructView::Schema::var::m_uint64, get_field_template<SimpleStructView::Schema::var::m_uint64>::type>
```
Comment (chadwalters) [inline cpp/test/core/container_extensibility.cpp]: Can you explain this part of the change?

File: CHANGELOG.md [hunk 4]
```diff
@@ -21,6 +21,9 @@ different versioning scheme, following the Haskell community's
 
 ### `gbc` and Bond compiler library ###
 
+* C++ codegen generates field reflection, that uses an explicit type 
+  derrived from reflection::FieldTemplate. This hides FieldTemplate 
+  details and, as result, shortens symbol names.
```
Comment (chadwalters) [inline CHANGELOG.md]: Minor: Change to "C++ codegen hides FieldTemplate details, shortening symbol names."

Comment (veselinp) [inline CHANGELOG.md]: Done.

File: cpp/test/core/container_extensibility.cpp (hunk 3 above)
Comment (veselinp) [inline cpp/test/core/container_extensibility.cpp]: I simplified this, and moved all logic into ordered_non_unique_field<>, please check cpp/test/core/multi_index.h

vc12 doesn't support constexp - ordered_non_unique_field requires member field address in compile time, which can be resolved via FieldTemplate.

**Comment by chwarr** [comment]
I will look again later today.

File: cpp/inc/bond/core/reflection.h [hunk 5]
```diff
@@ -101,24 +101,21 @@ struct FieldTemplate
     /// @brief Static data member describing field metadata
     static const Metadata& metadata;
 
-    /// @brief Static data member representing the field pointer
-    static const field_pointer field;
-
```
Comment (chwarr) [inline cpp/inc/bond/core/reflection.h]: This looks like a breaking change (removal of public member of public interface). Is it intentional? If so, what's motivating making this breaking change?

It appears that the replacement for this functionality is `&theFieldType.GetVariable(someStruct)`.

File: python/inc/bond/python/struct.h [hunk 6]
```diff
@@ -26,14 +26,14 @@ class def_readwrite_property
         : c(c)
     {}
 
-    template <typename Field>
-    void operator()(const Field&) const
+    template <uint16_t field_id, typename ModifierTag, typename Struct, typename FieldType, FieldType Struct::*field_ptr, const bond::Metadata* metadata_ptr>
```
Comment (chwarr) [inline python/inc/bond/python/struct.h]: Was this change required to get this to build?

* If so, then the introduction of the anonymous struct is not working as I'd expect it to (the anonymous struct should have an _is a_ relationship with `FieldTemplate`), and is a large breaking change.
* If this was just done to clean up the code, that's fine, but I need to get confirmation from you.

File: cpp/inc/bond/core/reflection.h (hunk 5 above)
Comment (veselinp) [inline cpp/inc/bond/core/reflection.h]: Good point, let me revisit this.

Comment (chwarr) [inline cpp/inc/bond/core/reflection.h]: If this materially reduces the lib and PDB size, then it's something we can consider. The next release for C++ will have some other breaking changes (albeit minor ones) like this.

File: python/inc/bond/python/struct.h (hunk 6 above)
Comment (veselinp) [inline python/inc/bond/python/struct.h]: This change is related to FieldTemplate<>::field change.

File: cpp/inc/bond/core/reflection.h (hunk 5 above)
Comment (veselinp) [inline cpp/inc/bond/core/reflection.h]: I added it back for now, we may want to handle vc12 explicitly to remove this symbol from generated .obj files.
